codeworld-compiler/src/CodeWorld/Compile/Requirements/Eval.hs

```diff
@@ -223,6 +224,46 @@ checkRule (NoWarningsExcept ex) = do
              let (SrcSpanInfo (SrcSpan _ l c _ _) _,_,x) = head warns
              failure $ "Warning found at line " ++ show l ++ ", column " ++ show c
 
+checkRule (TypeSignatures b) = withParsedCode $ \m -> do
+    let defs = nub $ topLevelNames m
+        noTypeSig = defs \\ (everything (++) (mkQ [] typeSigNames) m)
+
+        typeSigNames :: Decl SrcSpanInfo -> [String]
+        typeSigNames (TypeSig _ l _) = nameString <$> l
+        typeSigNames _ = []
+
+        nameString :: Name SrcSpanInfo -> String
+        nameString (Ident _ s) = s
+        nameString (Symbol _ s) = s
+
+    if | null noTypeSig || not b -> success
+       | otherwise -> failure $ "The declaration of `" ++ head noTypeSig
+           ++ "` has no type signature."
+
+checkRule (Blacklist bl) = withParsedCode $ \m -> do
```

Comment:
This won't look at the scope, so (for example) if you ban use of `red` in the blacklist, this will also catch:

    pic = colored(solidCircle(1), red)
      where red = RGB(1,0,0)

Ideally, that would still be allowed.

That said, though, fixing this will be much easier in a source plugin, where you can run after renaming.  So maybe this is okay as an interim step.

Reply:
I've had a look at trying to fix this, but I think you're right and it'll be easier to do later on. I've left a note about it in RequirementsChecker.md so we don't forget.